### ISSUE TITLE
Global Console - phase 1

### ISSUE THREAD

**Issue by k-wall** [EnMasseProject/enmasse#2408]
This design proposal is moved to: #2477

**Overview**

Currently in EnMasse a management console is co-deployed with each addresspace.  This gives the user the ability to manage the contents of the addressspace, but there is no UI that gives the user a view of the whole system nor the ability to add or delete addressspaces.  This is a gap.

We desire a Global Console that has the ability to view all addressspace that exist cluster wide, gives the ability to create new addressspaces, as well as update and delete existing ones.  In the longer term, it is envisaged that this new console will incorporate the address managing capabilities provided by the existing Console (and other features), but the first phase the global console will simple link to the existing consoles provided by the addressspace.

In addition it is desired that the Global Console is extensible to manage other Custom Resources too.  The use-case here is to manage the Custom Resources that correspond to other messaging infrastructure for instance Apache Strimzi.    Th architecture will allow the view and the control of the Custom Resource to be plugged in.  This will allow the UI to present a multi-stage create workflow targeted to the Resource in question.

In the long term, the Global Console will also present key statistics about each managed resource. This will allow the user to understand a glance the state of the whole system.  The statistics being presented will vary depending on the resource in question.   Statistic won't be included in the first phase.

**Design**

There will be a Global Console UI served from a new Global Console Server component serving the static content.

Authentication will be delegated to OAUTH.  In OpenShift this will be automatically configured at deployment to use the built OAUTH infrastructure.  On Kubernettes, the user will need to integrate their Kubernettes with a OAUTH provider.   

The UI will use the Kubernettes API to list and control the resources that it needs to expose for Management.  This part of the UI will be extensible so that the ability to control other resources can be enabled.  In the first the phase the two resource types exposed for management will be EnMasse's standard addressspace and brokered addressspace.   The UI will send the Bearer token in order that the server side may authorise the requests. 

Initially the UI will periodically poll the API for the list of resources (standard addressspace and brokered addressspace).  Later this will evolve to utilise watch for a more responsive UI with fewer overheads.

There will be support for a multi page create workflow that will allow the user to create new instances of managed resources.  The HTML and JS supporting this workflow will be implementations of a pluggable API exposed by the Console itself.

It will be possible for the user to click through from the the managed resource to the existing Console of that resource.  To enable this, the resource will expose a console URL.  EnMasse's Console will be improved to allow back navigation.   Authentication will be seem-less as the existing OAUTH session will be valid for the old Console too.

The mast head of the application will present a summary of the number of resources of each type current visible to the user, cluster-wide.  In the first version, it will present only the number of brokered and standard addressspaces, separately.

There are no new authorisation requirements.  Existing RBAC controls on managed resources will be sufficient to protect access.

The Global Console will use PatternFly 4.  The Global Console Server will be implemented in Go.  This will be responsible for serving the static content and a API endpoint returning the URLs for OAUTH redirect and the Kubernettes API endpoint.

The Global Console will be installed by the EnMasse Operator.

**Tasklist**

If/when you can, enumerate the tasks required to complete the feature in order to track progress as the feature is implemented.

**Comment by grs** [comment]
Will the component listing be by namespace?

**Comment by k-wall** [comment]
The console will list all address spaces from all namespaces that the user
has permissions to see. I think it makes sense for the names to be
qualified by namespace.

**Comment by grs** [comment]
An alternative would be to have a namespace/project selector on the console that indicates the current namespace/project, and show the relevant resources in that. (That is similar to how the openshift console works so would be familiar in a sense).

**Comment by k-wall** [comment]
> An alternative would be to have a namespace/project selector on the console that indicates the current namespace/project, and show the relevant resources in that. (That is similar to how the openshift console works so would be familiar in a sense).

Yes, reusing that approach from the OpenShift Console does seem appealing. Screen space might be an issue with the namespace prefixing.  I'll make sure this idea is considered.

**Comment by k-wall** [comment]
Regarding OAuth2 integration, the Global Console will be a Single Page App (SPA).  [RFC 6749 section 1.3.2](https://tools.ietf.org/html/rfc6749#section-1.3.2) recommends the OAuth2 implict workflow for such applications (this is the workflow where the response from the authorisation endpoint returns an access token rather than authorisation code).

However, there is a new [best practice recommendation ](https://auth0.com/blog/oauth2-implicit-grant-and-spa/) in the form of a [Draft RFC](url
https://tools.ietf.org/html/draft-ietf-oauth-security-topics-11) that recommends a different approach leveraging on PKCE for client type `public`. (nice explainer: https://developer.okta.com/authentication-guide/auth-overview/#authorization-code-with-pkce-flow)

I don't know how will the new recommendation will be supported by existing public OAuth providers (that might be used by an EnMasse user on Kubernettes).  Nor have I tried the approach on OpenShift.

I think we should initially use `implicit` workflow and then look at the improved approach at a later point.

**Comment by k-wall** [comment]
> > An alternative would be to have a namespace/project selector on the console that indicates the current namespace/project, and show the relevant resources in that. (That is similar to how the openshift console works so would be familiar in a sense).
> 
> Yes, reusing that approach from the OpenShift Console does seem appealing. Screen space might be an issue with the namespace prefixing. I'll make sure this idea is considered.

@grs it turns out the wireframes actually have provision for filtering by namespace already.

**Comment by k-wall** [comment]
The above plan envisaged the client side HTML?JS application calling out to the Kubernetes API server falls foul of the default CORS restrictions imposed by OpenShift/Kubernetes API Server [1].   It is possible to specify additional CORS hosts names into the [whitelist](https://docs.openshift.com/container-platform/3.11/install_config/master_node_configuration.html#master-config-asset-config) specified in the master's configure (in Kubernettes https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#enable-cors) . However, we probably won't want this approach, as additional administrative burden is an impediment.

To avoid the need for the reconfiguration, we could write our own proxy that proxies the requests through to the api-server.  In order to support watches, the implementation would need to be capable of HTTP 1.1 only and support chunked requests (aka HTTP streams).

To avoid writing our own code we could reuse an implementation.  We could investigate the practicalities of reusing https://github.com/openshift/origin-web-console-server/tree/enterprise-3.11 wholesale.  The other possibility is to employ Apache HTTPD as an reverse proxy.

[1] https://blog.openshift.com/kubernetes-deep-dive-api-server-part-1/

**Comment by k-wall** [comment]
Regarding calling out to API server from the HTML/JS app, I investigated option 2).  It turns out that `mod_auth_openidc` can't help. The plugin mainly targets OpenID Connect use-cases in addition to some plain OAuth2 use-cases.  Unfortunately, it does not support the OAuth2 authorisation code grant login workflow, which is the one I wanted.   Proxying the API server with Apache HTTPD with `ProxyPass`, so a combination of HTML/JS app using the implict OAuth2 workflow with HTTPD serving static content and proxying requests to the API server is viable.

I also investigated an alternative approach of using [openshift oauth-proxy](https://github.com/openshift/oauth-proxy).     It provides:

> A reverse proxy and static file server that provides authentication and authorization to an OpenShift OAuth server or Kubernetes master supporting the 1.6+ remote authorization endpoints to validate access to content.

This approach works with one wrinkle.  When proxying requests down to the API server, oauth-proxy sends the authorisation token belonging to the autenticate user through with `X-Forward-For` style header.  The Kubernetes API does not understand this header, so it treats the request as unauthenticated.   It is possible to workaround this problem either by using a Apache HTTPD sidecar that simply rewrites the request header (`SetEnvIf`/`RequestHeader` directives).  Longer term, It would also be simple to enhance oauth-proxy to have the ability to pass the Authorization header through to the backend.
